### DISCUSSION TITLE
Camera to the world

### DISCUSSION THREAD

**Discussion by Daoudou** [openscenegraph/OpenSceneGraph#1208] [Q&A]
Hello.

I have another problem. I use setViewMatrix in my camera to have a view in my world but how can i do to enter into the world ?
if i what to in my skydome the camera stay outside the skydome.

If i don't mistake placing the camera at a certain point P applies a translation of -P to the whole world.
Or i have to change the projection like it's like the lens of a camera.

    camera->setViewport(new osg::Viewport(iViewport[0], iViewport[1], iViewport[2], iViewport[3]));
    camera->setProjectionMatrix(projectionMatrix);
    camera->setViewMatrix(modelViewMatrix);

![Capture](https://user-images.githubusercontent.com/98310378/223147168-aae21a90-328b-4edd-8c55-6d93d03d5f8d.PNG)

Thanks in advance.

SOUFFOU

**Comment by trajcenikolovnick** [comment]
try
camera->setViewMatrixAsLookAt( eye, center, up )

On Mon, Mar 6, 2023 at 5:01 PM Daoudou ***@***.***> wrote:

> Hello.
>
> I have another problem. I use setViewMatrix in my camera to have a view in
> my world but how can i do to enter in the world ?
> if i what to in my skydome the camera stay outside the skydome.
>
> If i don't mistake placing the camera at a certain point P applies a
> translation of -P to the whole world.
>
> camera->setViewport(new osg::Viewport(iViewport[0], iViewport[1],
> iViewport[2], iViewport[3]));
> camera->setProjectionMatrix(projectionMatrix);
> camera->setViewMatrix(modelViewMatrix);
>
> [image: Capture]
> <https://user-images.githubusercontent.com/98310378/223147168-aae21a90-328b-4edd-8c55-6d93d03d5f8d.PNG>
>
> Thanks in advance.
>
> SOUFFOU
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openscenegraph/OpenSceneGraph/discussions/1208>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC7ND7EU76I2GX3YL5U7XQLW2X34RANCNFSM6AAAAAAVRHRNUI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
trajce nikolov nick

**Comment by Daoudou** [comment]
Okay i will try it thanks

I will also see if I can associate its with the modelViewMatrix that I get from my OpenGL project to have the same view dynamically.

**Comment by Daoudou** [comment]
Problem solved 👍 

I get the modelViewMatrix and the projectionMatrix of opengl in OpenSceneGraph

            osg::Matrixd modelViewMatrix(modelMatrix[0], modelMatrix[1], modelMatrix[2], modelMatrix[3],
            modelMatrix[4], modelMatrix[5], modelMatrix[6], modelMatrix[7],
            modelMatrix[8], modelMatrix[9], modelMatrix[10], modelMatrix[11],
            modelMatrix[12], modelMatrix[13], modelMatrix[14], modelMatrix[15]
         );

        osg::Matrixd projectionMatrix(projetMatrix[0], projetMatrix[1], projetMatrix[2], projetMatrix[3],
            projetMatrix[4], projetMatrix[5], projetMatrix[6], projetMatrix[7],
            projetMatrix[8], projetMatrix[9], projetMatrix[10], projetMatrix[11],
            projetMatrix[12], projetMatrix[13], projetMatrix[14], projetMatrix[15]
        );

I create a first camera and i set it in the first camera

        Camera->setViewMatrix(modelViewMatrix);
        Camera->setProjectionMatrix(projectionMatrix);
        Camera->setViewport(new osg::Viewport(iViewport[0], iViewport[1], iViewport[2], iViewport[3]));

Get the LookAt matrix

        Camera->getViewMatrixAsLookAt(eye, center, up);
 
 Get the camera of my viewer and calulate the position of the camera with the LookAt of the first camera 
 
        eye1 = eye;
        center1 = center;
        up1 = up;

        eye1.x() = (((eye.x() + (camera_pos[0] / 2.)))*1.5) / ratioZoom;
        center1.x() = (((center.x() + ((-camera_pos[0]) / 2)))*1.5) * ratioZoom;
        
        eye1.y() = (((eye.y() + (camera_pos[1]/2.)))*1.5) / ratioZoom;
        center1.y() = (((center.y() + ((-camera_pos[1]) / 2)))*1.5) * ratioZoom;

        eye1.z() = (((eye.z() + (camera_pos[2] / 2)))*1.5) / ratioZoom;
        center1.z() = (((center.z() + ((-camera_pos[2])/2)))*1.5) * ratioZoom;
        
(Yes, i take my head a lot )
 i set this here if someone want to do the same thing

Thanks.
